lib/bandiera/caching_feature_service.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'lru_redux'
+
+module Bandiera
+  class CachingFeatureService < SimpleDelegator
+    CACHE_SIZE = 100
+    CACHE_TTL  = 10
+
+    attr_reader :cache
+
+    def initialize(delegate)
+      @cache = LruRedux::TTL::ThreadSafeCache.new(CACHE_SIZE, CACHE_TTL)
+
+      super
+    end
+
+    def find_group(name)
+      cache.getset(:"find_group:#{name}") { super }
+    end
+
+    def fetch_groups
+      cache.getset(:fetch_groups) { super }
+    end
+
+    def fetch_group_features(group)
+      cache.getset(:"fetch_group_features:#{group}") { super }
+    end
+
+    def fetch_feature(group, feature_name)
+      cache.getset(:"fetch_feature:#{group}:#{feature_name}") { super }
+    end
+
+    def add_group(group)
+      cache.delete(:"find_group:#{group}")
+      cache.delete(:fetch_groups)
+
+      super
+    end
+
+    def add_feature(data)
+      cache.delete(:"fetch_feature:#{data[:group]}:#{data[:name]}")
+      cache.delete(:"fetch_group_features:#{data[:group]}")
+
+      super
+    end
+
+    def add_features(features)
+      features.each do |data|
+        cache.delete(:"fetch_feature:#{data[:group]}:#{data[:name]}")
+        cache.delete(:"fetch_group_features:#{data[:group]}")
+      end
+
+      super
+    end
+
+    def remove_feature(group, feature_name)
+      cache.delete(:"fetch_feature:#{group}:#{feature_name}")
+      cache.delete(:"fetch_group_features:#{group}")
+
+      super
+    end
+
+    def update_feature(group, feature_name, feature_data)
+      cache.delete(:"fetch_feature:#{group}:#{feature_name}")
+      cache.delete(:"fetch_group_features:#{group}")
+
+      super
+    end
```

Comment:
Whenever I create classes with the same API (they both play the same role) I would write `shared_examples`, at a minimum ensuring they respond to the same methods so that I can insert in both specs (in this case) `it_behaves_like 'a feature service'`

Then if you want to extend what that role does, you have an obvious spec to change to drive the code changes

Can we do that here ? I can contribute the code if you like.

Reply:
That's a bloody good idea Andy. Yep, if you've got the time that would be fantastic to do, otherwise I can take a look next week. :)

Reply:
Done.